yt/yt/tests/integration/controller/test_vanilla_operation.py

```diff
@@ -2764,3 +2764,475 @@ def test_decrease_job_count(self):
+
+    @authors("pavel-bash")
+    def test_restart_always_after_failure(self):
+        """
+        Check the {RestartPolicy=Always & sidecar-finished-with-failure} case; we expect that
+        the sidecar will be launched for the 2nd time, emitting the corresponding event.
+        """
+        master_command = " ; ".join(
+            [
+                events_on_fs().notify_event_cmd("master_job_started_${YT_JOB_COOKIE}"),
+                events_on_fs().wait_event_cmd("finish"),
+            ]
+        )
+        sidecar_command = f"""
+#!/bin/bash
+if [ -f {events_on_fs()._get_event_filename("sidecar_first_job_started")} ]; then
+  # It's the 2nd time the sidecar is launched
+  {events_on_fs().notify_event_cmd("sidecar_second_job_started")}
+  {events_on_fs().wait_event_cmd("finish_sidecar_second")}
+else
+  {events_on_fs().notify_event_cmd("sidecar_first_job_started")}
+  {events_on_fs().wait_event_cmd("finish_sidecar_first")}
+  exit 1
+fi
+"""
+
+        op = self.launch_operation(master_command, sidecar_command, "always")
+
+        events_on_fs().wait_event("master_job_started_0", timeout=datetime.timedelta(1000))
+
+        events_on_fs().wait_event("sidecar_first_job_started", timeout=datetime.timedelta(1000))
+        events_on_fs().notify_event("finish_sidecar_first")
+
+        events_on_fs().wait_event("sidecar_second_job_started", timeout=datetime.timedelta(1000))
+        events_on_fs().notify_event("finish_sidecar_second")
+
+        events_on_fs().notify_event("finish")
+
+        self.ensure_operation_finish(op)
+
+    @authors("pavel-bash")
+    def test_restart_on_failure_after_success(self):
+        """
+        Check the {RestartPolicy=OnFailure & sidecar-finished-with-success} case; the sidecar
+        must not be restarted.
+        """
+        master_command = " ; ".join(
+            [
+                events_on_fs().notify_event_cmd("master_job_started_${YT_JOB_COOKIE}"),
+                events_on_fs().wait_event_cmd("finish"),
+            ]
+        )
+        sidecar_command = f"""
+#!/bin/bash
+if [ -f {events_on_fs()._get_event_filename("sidecar_first_job_started")} ]; then
+  {events_on_fs().notify_event_cmd("sidecar_second_job_started")}
+else
+  {events_on_fs().notify_event_cmd("sidecar_first_job_started")}
+  {events_on_fs().wait_event_cmd("finish_sidecar_first")}
+fi
+"""
+
+        op = self.launch_operation(master_command, sidecar_command, "on_failure")
+
+        events_on_fs().wait_event("master_job_started_0", timeout=datetime.timedelta(1000))
+
+        events_on_fs().wait_event("sidecar_first_job_started", timeout=datetime.timedelta(1000))
+        events_on_fs().notify_event("finish_sidecar_first")
+
+        # Wait here a little bit to allow the sidecar to start (if it's going to)
+        time.sleep(5)
+
+        events_on_fs().notify_event("finish")
+
+        self.ensure_operation_finish(op)
+
+        assert not events_on_fs().check_event("sidecar_second_job_started")
+
```

Comment:
These tests are somewhat verbose/hard to grasp, can we try to add either some more detailed scenario description in comments, or comments throughout the test to help with that? If possible.